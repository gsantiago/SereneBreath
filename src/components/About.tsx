import React from "react";
import { Popover } from "./Popover";

export function About({ isVisible }) {
  return (
    <Popover isVisible={isVisible} position="bottom">
      <section className="p-3">
        <p>made with love by @gsantiago</p>
      </section>
    </Popover>
  );
}
