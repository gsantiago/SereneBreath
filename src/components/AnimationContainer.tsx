import { animated, SpringValue, useSpringValue } from "@react-spring/web";

export const SIZE = 200;

export interface AnimationContainerProps {
  scale?: SpringValue<number>;
  children: React.ReactNode;
}

export function AnimationContainer({
  scale,
  children,
}: AnimationContainerProps) {
  const defaultScale = useSpringValue(1);

  return (
    <animated.div
      className="flex items-center justify-center "
      style={{ width: SIZE, height: SIZE, scale: scale ?? defaultScale }}
    >
      {children}
    </animated.div>
  );
}
