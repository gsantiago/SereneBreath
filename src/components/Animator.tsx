import React, { useEffect, useState } from "react";
import { AnimationProps } from "@/config/types";

export interface AnimatorProps
  extends Omit<AnimationProps, "state" | "isHolding"> {
  animation: React.FC<AnimationProps>;
}

export function Animator({
  animation: Animation,
  currentStep,
  pattern,
}: AnimatorProps) {
  const [state, setState] = useState<AnimationProps["state"]>("idle");

  const isHolding = currentStep === 1 || currentStep === 3;

  useEffect(() => {
    if (currentStep === 0 || currentStep === 1) {
      setState("inhaling");
    } else {
      setState("exhaling");
    }
  }, [currentStep]);

  return (
    <Animation
      currentStep={currentStep}
      pattern={pattern}
      state={state}
      isHolding={isHolding}
    />
  );
}
