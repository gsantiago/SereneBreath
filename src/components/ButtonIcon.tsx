import React from "react";

export interface ButtonIconProps
  extends Omit<React.HTMLProps<HTMLButtonElement>, "type"> {
  isActive?: boolean;
  position?: "top" | "bottom";
}

export function ButtonIcon({
  isActive,
  position = "top",
  ...props
}: ButtonIconProps) {
  const positionClass =
    position === "top" ? "rounded-t-md" : "rounded-b-md shadow-md";

  return (
    <button
      type="button"
      className={`px-2 pb-2 pt-2 text-3xl dark:text-white ${
        isActive
          ? `z-20 bg-white dark:bg-slate-800 ${positionClass}`
          : "rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
      }`}
      {...props}
    />
  );
}
