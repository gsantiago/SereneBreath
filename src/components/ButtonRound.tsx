import React, { HTMLAttributes } from "react";

export function ButtonRound({
  style,
  ...props
}: HTMLAttributes<HTMLButtonElement>) {
  return (
    <button
      className="flex h-12 w-12 items-center justify-center rounded-full bg-teal-500 text-2xl text-white hover:bg-teal-600"
      {...props}
    />
  );
}
