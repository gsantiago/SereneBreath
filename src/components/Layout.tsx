import React, { useState } from "react";

import { Header } from "@/components/Header";

export interface LayoutProps {
  showHeader: boolean;
  children: React.ReactNode;
}

export function Layout({ showHeader, children }: LayoutProps) {
  const [showSettings, setShowSettings] = useState(false);

  return (
    <div className="h-full  bg-sky-100 dark:bg-slate-900">
      <div className="relative m-auto flex h-full flex-col items-center justify-between overflow-hidden ">
        <Header
          isVisible={showHeader}
          showSettings={showSettings}
          onToggleSettings={() => {
            setShowSettings((show) => !show);
          }}
        />
        <div className="flex h-full w-full items-end justify-center pb-5 md:items-center">
          {children}
        </div>
      </div>
      {showSettings && (
        <div
          onClick={() => {
            setShowSettings(false);
          }}
          className="absolute left-0 top-0 h-full w-full"
        />
      )}
    </div>
  );
}
