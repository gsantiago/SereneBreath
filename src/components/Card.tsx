import { RingAnimation } from "@/animations/RingAnimation";
import { Animator } from "@/components/Animator";
import { Pattern } from "@/config/types";

export const WIDTH = 250;
const HEIGHT = "100%";

export interface CardProps {
  title: string;
  description: string;
  pattern: Pattern;
}

export function Card({ title, description, pattern }: CardProps) {
  return (
    <section
      className="overflow-hidden rounded-lg bg-white"
      style={{ width: WIDTH, height: HEIGHT, paddingBottom: 45 }}
    >
      <div className="flex h-1/2 w-full items-center justify-center bg-orange-100">
        <div style={{ transform: "scale(0.8)" }}>
          <Animator
            currentStep={3}
            pattern={[2, 2, 2, 2]}
            animation={RingAnimation}
          />
        </div>
      </div>
      <div className="p-5">
        <h1 className="text-xl font-bold sm:text-2xl">{title}</h1>
        <p className="my-1 text-xs text-gray-400">{pattern.join("-")}</p>
        <p className="text-sm text-gray-800 sm:text-base">{description}</p>
      </div>
    </section>
  );
}
