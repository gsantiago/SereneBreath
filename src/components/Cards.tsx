import { Carousel } from "@/components/Carousel";
import { Card, WIDTH as CARD_WIDTH } from "@/components/Card";

import { useTranslation } from "@/hooks/useTranslation";
import { useStorage } from "@/hooks/useStorage";

import { techniques } from "@/config/techniques";
import { Technique } from "@/config/types";

export function Cards() {
  const { t } = useTranslation();

  const [technique, setTechnique] = useStorage("technique");
  const activeIndex = techniques.findIndex((t) => t.name === technique);

  return (
    <Carousel<Technique>
      activeIndex={activeIndex}
      itemWidth={CARD_WIDTH}
      items={[...techniques]}
      onChange={(item) => setTechnique(item.name)}
    >
      {({ item, isActive }) => (
        <Card
          title={t(`techniques.${item.name}.title`)}
          description={t(`techniques.${item.name}.description`)}
          pattern={item.pattern}
          isActive={isActive}
        />
      )}
    </Carousel>
  );
}
