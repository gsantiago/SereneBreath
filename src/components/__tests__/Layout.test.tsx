import { render, axe, screen } from "@/test-utils";
import { Layout, LayoutProps } from "@/components/Layout";

const setup = (props?: Partial<LayoutProps>) =>
  render(
    <Layout showHeaderAndFooter {...props}>
      content
    </Layout>
  );

test("renders successfuly", () => {
  const { container } = setup();
  expect(container).toMatchSnapshot();
});

test("accessibility", async () => {
  const { container } = setup();
  expect(await axe(container)).toHaveNoViolations();
});

test("does not show header and footer", () => {
  setup({ showHeaderAndFooter: false });

  expect(screen.getByTestId("header")).not.toBeVisible();
  expect(screen.getByTestId("footer")).toHaveClass("opacity-0");
});
