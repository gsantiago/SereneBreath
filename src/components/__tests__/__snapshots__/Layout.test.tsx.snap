// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders successfuly 1`] = `
<div>
  <div
    class=" h-full  bg-sky-100 dark:bg-slate-900"
  >
    <div
      class="relative m-auto flex h-full max-w-5xl flex-col items-center justify-between p-5"
    >
      <header
        class="relative flex w-full items-center justify-between"
        data-testid="header"
        style="opacity: 0; transform: none;"
      >
        <h1
          class="text-2xl font-bold dark:text-white"
        >
          SereneBreath
        </h1>
        <div
          class="z-10 flex"
        >
          <button
            class="px-2 pb-2 pt-2 text-3xl dark:text-white rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
            title="Settings"
            type="button"
          >
            <svg
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.343 3.94c.09-.542.56-.94 1.11-.94h1.093c.55 0 1.02.398 1.11.94l.149.894c.07.424.384.764.78.93.398.164.855.142 1.205-.108l.737-.527a1.125 1.125 0 011.45.12l.773.774c.39.389.44 1.002.12 1.45l-.527.737c-.25.35-.272.806-.107 1.204.165.397.505.71.93.78l.893.15c.543.09.94.56.94 1.109v1.094c0 .55-.397 1.02-.94 1.11l-.893.149c-.425.07-.765.383-.93.78-.165.398-.143.854.107 1.204l.527.738c.32.447.269 1.06-.12 1.45l-.774.773a1.125 1.125 0 01-1.449.12l-.738-.527c-.35-.25-.806-.272-1.203-.107-.397.165-.71.505-.781.929l-.149.894c-.09.542-.56.94-1.11.94h-1.094c-.55 0-1.019-.398-1.11-.94l-.148-.894c-.071-.424-.384-.764-.781-.93-.398-.164-.854-.142-1.204.108l-.738.527c-.447.32-1.06.269-1.45-.12l-.773-.774a1.125 1.125 0 01-.12-1.45l.527-.737c.25-.35.273-.806.108-1.204-.165-.397-.505-.71-.93-.78l-.894-.15c-.542-.09-.94-.56-.94-1.109v-1.094c0-.55.398-1.02.94-1.11l.894-.149c.424-.07.765-.383.93-.78.165-.398.143-.854-.107-1.204l-.527-.738a1.125 1.125 0 01.12-1.45l.773-.773a1.125 1.125 0 011.45-.12l.737.527c.35.25.807.272 1.204.107.397-.165.71-.505.78-.929l.15-.894z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="absolute  w-full rounded-md  bg-white py-2 shadow-md dark:bg-slate-800 sm:w-80 hidden right-0 top-full rounded-tr-none z-10"
          >
            <div
              class="px-5 py-2"
            >
              <div
                class="grid grid-cols-6 items-end"
              >
                <div
                  class="col-span-4"
                >
                  <label
                    class="text-sm font-medium text-gray-800 dark:text-white"
                    for="technique"
                  >
                    Technique
                  </label>
                </div>
                <div
                  class="col-span-2 text-right"
                />
              </div>
              <div
                class="mt-2"
              >
                <select
                  class="w-full cursor-pointer rounded-md border-gray-300 text-sm dark:text-gray-700"
                  id="technique"
                >
                  <option
                    value="Square"
                  >
                    Square (4-4-4-4)
                  </option>
                  <option
                    value="Awake"
                  >
                    Awake (6-0-2-0)
                  </option>
                  <option
                    value="Deep Calm"
                  >
                    Deep Calm (4-7-8-0)
                  </option>
                  <option
                    value="Pranayama"
                  >
                    Pranayama (7-4-8-4)
                  </option>
                  <option
                    value="Ujjayi"
                  >
                    Ujjayi (7-0-7-0)
                  </option>
                </select>
              </div>
            </div>
            <div
              class="px-5 py-2"
            >
              <div
                class="grid grid-cols-6 items-end"
              >
                <div
                  class="col-span-4"
                >
                  <label
                    class="text-sm font-medium text-gray-800 dark:text-white"
                    for="time"
                  >
                    Time
                  </label>
                  <p
                    class="text-xs text-gray-500 dark:text-gray-400"
                  >
                    Exercise time in minutes
                  </p>
                </div>
                <div
                  class="col-span-2 text-right"
                >
                  <div
                    class="inline-flex items-center"
                  >
                    <button
                      class="flex h-6 w-6 items-center justify-center rounded-md bg-blue-600 text-white hover:bg-blue-700"
                      title="Decrease"
                      type="button"
                    >
                      <svg
                        class="h-5 w-5"
                        fill="currentColor"
                        viewBox="0 0 20 20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M6.75 9.25a.75.75 0 000 1.5h6.5a.75.75 0 000-1.5h-6.5z"
                        />
                      </svg>
                    </button>
                    <p
                      class="mx-2 w-4 text-center text-sm dark:text-white"
                    >
                      1
                    </p>
                    <button
                      class="flex h-6 w-6 items-center justify-center rounded-md bg-blue-600 text-white hover:bg-blue-700"
                      title="Increase"
                      type="button"
                    >
                      <svg
                        class="h-5 w-5"
                        fill="currentColor"
                        viewBox="0 0 20 20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M10.75 6.75a.75.75 0 00-1.5 0v2.5h-2.5a.75.75 0 000 1.5h2.5v2.5a.75.75 0 001.5 0v-2.5h2.5a.75.75 0 000-1.5h-2.5v-2.5z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="px-5 py-2"
            >
              <div
                class="grid grid-cols-6 items-end"
              >
                <div
                  class="col-span-4"
                >
                  <label
                    class="text-sm font-medium text-gray-800 dark:text-white"
                    for="guide"
                  >
                    Guide Breath
                  </label>
                  <p
                    class="text-xs text-gray-500 dark:text-gray-400"
                  >
                    Select your favorite guide
                  </p>
                </div>
                <div
                  class="col-span-2 text-right"
                >
                  <div
                    class="relative inline-flex cursor-pointer items-center"
                  >
                    <input
                      checked=""
                      class="peer sr-only"
                      id="guide"
                      type="checkbox"
                    />
                    <button
                      class="peer h-6 w-11 rounded-full bg-gray-200 after:absolute after:top-[2px] after:left-[2px] after:h-5 after:w-5 after:rounded-full after:border after:border-gray-300 after:bg-white after:transition-all after:content-[''] peer-checked:bg-blue-600 peer-checked:after:translate-x-full peer-checked:after:border-white peer-focus:outline-none peer-focus:ring-4 peer-focus:ring-blue-300 dark:border-gray-600 dark:bg-gray-700 dark:peer-focus:ring-blue-800"
                      title="Switch"
                    />
                  </div>
                </div>
              </div>
              <div
                class="mt-2"
              >
                <div
                  class="flex items-center"
                >
                  <input
                    checked=""
                    class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
                    id="option-female"
                    name="guide"
                    type="radio"
                    value="female"
                  />
                  <label
                    class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
                    for="option-female"
                  >
                    Female voice
                  </label>
                </div>
                <div
                  class="flex items-center"
                >
                  <input
                    class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
                    id="option-male"
                    name="guide"
                    type="radio"
                    value="male"
                  />
                  <label
                    class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
                    for="option-male"
                  >
                    Male voice
                  </label>
                </div>
                <div
                  class="flex items-center"
                >
                  <input
                    class="h-4 w-4 cursor-pointer border-gray-300 text-blue-600 focus:ring-blue-600"
                    id="option-bell"
                    name="guide"
                    type="radio"
                    value="bell"
                  />
                  <label
                    class="text-md block cursor-pointer pl-2 text-xs leading-6 text-gray-700 dark:text-gray-100"
                    for="option-bell"
                  >
                    Bell
                  </label>
                </div>
              </div>
            </div>
            <div
              class="px-5 py-2"
            >
              <div
                class="grid grid-cols-6 items-end"
              >
                <div
                  class="col-span-4"
                >
                  <label
                    class="text-sm font-medium text-gray-800 dark:text-white"
                    for="theme"
                  >
                    Theme
                  </label>
                </div>
                <div
                  class="col-span-2 text-right"
                />
              </div>
              <div
                class="mt-2"
              >
                <div
                  class="flex"
                >
                  <label
                    class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-blue-600 bg-blue-600 text-white"
                  >
                    <input
                      checked=""
                      class="sr-only"
                      name="theme"
                      type="radio"
                      value="system"
                    />
                    System
                  </label>
                  <label
                    class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-gray-300 bg-gray-50"
                  >
                    <input
                      class="sr-only"
                      name="theme"
                      type="radio"
                      value="dark"
                    />
                    Dark
                  </label>
                  <label
                    class="cursor-pointer border  py-2 px-3 text-xs first:rounded-tl-md first:rounded-bl-md first:border-r-0 last:rounded-tr-md last:rounded-br-md last:border-l-0 hover:border-blue-600 hover:bg-blue-600 hover:text-white border-gray-300 bg-gray-50"
                  >
                    <input
                      class="sr-only"
                      name="theme"
                      type="radio"
                      value="light"
                    />
                    Light
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </header>
      <div
        class="flex h-full w-full items-end justify-center md:items-center"
      >
        content
      </div>
      <footer
        class="relative flex w-full justify-between text-xs transition-opacity dark:text-white opacity-1"
        data-testid="footer"
      >
        <div
          class="w-40"
        >
          <div>
            <label
              class="sr-only"
              for="locale"
            >
              Language
            </label>
            <select
              class="w-full cursor-pointer rounded-md border-gray-300 text-sm dark:text-gray-700"
              id="locale"
            >
              <option
                value="en"
              >
                English
              </option>
              <option
                value="pt"
              >
                Português
              </option>
            </select>
          </div>
        </div>
        <button
          class="px-2 pb-2 pt-2 text-3xl dark:text-white rounded-md hover:bg-white hover:shadow-sm dark:hover:bg-slate-800"
          title="About"
          type="button"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.25 11.25l.041-.02a.75.75 0 011.063.852l-.708 2.836a.75.75 0 001.063.853l.041-.021M21 12a9 9 0 11-18 0 9 9 0 0118 0zm-9-3.75h.008v.008H12V8.25z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
        <div
          class="absolute  w-full rounded-md  bg-white py-2 shadow-md dark:bg-slate-800 sm:w-80 hidden right-0 bottom-full rounded-br-none z-10"
        >
          <section
            class="px-3 py-1"
          >
            <h1
              class="text-lg font-bold"
            >
              SereneBreath
            </h1>
            <p
              class="mt-2 text-xs"
            >
              SereneBreath.com is a small webapp to help you improve your breathing, relaxing and managing your stress.
            </p>
            <p
              class="mt-2 text-xs"
            >
              This project is open source and open for new contributions. Check out its repo for more info:
               
              <a
                class="text-blue-600 underline dark:text-sky-300"
                href="https://github.com/gsantiago/SereneBreath"
                target="_blank"
              >
                Github Repo.
              </a>
            </p>
            <p
              class="mt-2 text-xs"
            >
              English and bell audio files are from the great Breathly app, which also inspired this app.
               
              <a
                class="text-blue-600 underline dark:text-sky-300"
                href="https://breathly.app/"
                target="_blank"
              >
                Breathly Site.
              </a>
            </p>
            <div
              class="mt-5 text-center"
            >
              <small
                class="text-xs text-gray-400"
              >
                v0.0.1
              </small>
            </div>
          </section>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
